controller/deployer/main.go

```diff
@@ -0,0 +1,142 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"os"
+
+	"github.com/flynn/flynn/Godeps/_workspace/src/github.com/bgentry/que-go"
+	"github.com/flynn/flynn/Godeps/_workspace/src/github.com/jackc/pgx"
+	"github.com/flynn/flynn/controller/client"
+	"github.com/flynn/flynn/controller/deployer/strategies"
+	ct "github.com/flynn/flynn/controller/types"
+	"github.com/flynn/flynn/pkg/postgres"
+	"github.com/flynn/flynn/pkg/shutdown"
+)
+
+type context struct {
+	db     *postgres.DB
+	client *controller.Client
+}
+
+func main() {
+	client, err := controller.NewClient("", os.Getenv("AUTH_KEY"))
+	if err != nil {
+		log.Fatalln("Unable to create controller client:", err)
+	}
+
+	postgres.Wait("")
+	db, err := postgres.Open("", "")
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	cxt := context{db: db, client: client}
+
+	pgxcfg, err := pgx.ParseURI(fmt.Sprintf("http://%s:%s@%s/%s", os.Getenv("PGUSER"), os.Getenv("PGPASSWORD"), db.Addr(), os.Getenv("PGDATABASE")))
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	pgxpool, err := pgx.NewConnPool(pgx.ConnPoolConfig{
+		ConnConfig:   pgxcfg,
+		AfterConnect: que.PrepareStatements,
+	})
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer pgxpool.Close()
+
+	q := que.NewClient(pgxpool)
+	wm := que.WorkMap{"deployment": cxt.HandleJob}
+
+	workers := que.NewWorkerPool(q, wm, 10)
+	go workers.Start()
+	shutdown.BeforeExit(func() { workers.Shutdown() })
+
+	<-make(chan bool) // block and keep running
+}
+
+func (c *context) HandleJob(job *que.Job) (e error) {
+	var args ct.DeployID
+	if err := json.Unmarshal(job.Args, &args); err != nil {
+		// TODO: log error
+		return err
+	}
+	deployment, err := c.client.GetDeployment(args.ID)
+	if err != nil {
+		// TODO: log error
+		return err
+	}
+	// for recovery purposes, fetch old formation
+	f, err := c.client.GetFormation(deployment.AppID, deployment.OldReleaseID)
+	if err != nil {
+		// TODO: log error
+		return err
+	}
+	strategyFunc, err := strategy.Get(deployment.Strategy)
+	if err != nil {
+		// TODO: log error
+		return err
+	}
+	events := make(chan ct.DeploymentEvent)
+	go func() {
+		for ev := range events {
+			ev.DeploymentID = deployment.ID
+			if err := c.createDeploymentEvent(ev); err != nil {
+				log.Print(err)
+			}
+		}
+	}()
+	defer func() {
+		// rollback failed deploy
+		if e != nil {
+			events <- ct.DeploymentEvent{
+				ReleaseID: deployment.NewReleaseID,
+				Status:    "failed",
+			}
+			if err = c.rollback(deployment, f); err != nil {
+				e = err
+			}
```

Comment:
I simplified it to that, but then I noticed that if rollback succeeds and returns `nil`, then we will return nil from the worker, and the job will be marked as `done` instead of `failed` (and the worker then won't retry the job)